Phoenix & Dove: init at 2025.02.21.1

jbgi opened the pull request:
[Phoenix](https://codeberg.org/celenity/Phoenix) and [Dove](https://codeberg.org/celenity/Dove) are suites of configurations & advanced modifications for Mozilla Firefox & Thunderbird (respectively) with a focus on privacy, security, freedom, & usability.

Fix #379289.

Usage:

```nix
  environment.etc."firefox/defaults/pref/phoenix-desktop.js".source =
    "${pkgs.phoenix}/prefs/phoenix-desktop.js";
  environment.etc."firefox/phoenix/userjs".source = "${pkgs.phoenix}/userjs";
  environment.etc."firefox/phoenix/configs".source = "${pkgs.phoenix}/configs";

  environment.etc."thunderbird/defaults/pref/dove.js".source =
    "${pkgs.dove}/prefs/dove.js";

  nixpkgs.overlays = [
    (self: super: {
      # Alternatively, use programs.firefox options, but this way also works for home-manager:
      firefox = super.firefox.override {
        extraPoliciesFiles = [ "${pkgs.phoenix}/policies.json" ];
        extraPrefsFiles = [ "${pkgs.phoenix}/phoenix.cfg" ];
      };

      thunderbird = super.thunderbird.override {
        extraPoliciesFiles = [ "${pkgs.dove}/policies.json" ];
        extraPrefsFiles = [ "${pkgs.dove}/dove.cfg" ];
      };
    })
  ];
```

WARNING: by default phoenix config will delete browsing history when closing firefox (got to setting to re-enabled if needed)

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jbgi committed:
phoenix: init at 2025.02.21.1

 Phoenix is a suite of configurations & advanced modifications for
 Mozilla Firefox with a focus on privacy, security, freedom, & usability.
 https://phoenix.celenity.dev/

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5191

jbgi committed:
dove: init at 2025.02.21.1

 Dove is a suite of configurations & advanced modifications for Mozilla Thunderbird,
 designed to put the user first - with a focus on privacy, security, freedom, & usability.
 https://dove.celenity.dev

linsui commented on pkgs/by-name/do/dove/package.nix:
```diff
@@ -0,0 +1,43 @@
+  pname = "dove";
+  version = "2025.02.01.1";
+
+  src = fetchzip {
```
Use fetchFromGitea.

jbgi replied:
this will fetch the whole repo with all the build scripts and more, whereas I'm only interested in the few files in the zip archive. Is that ok?

linsui replied:
It also fetches the zip.

jbgi replied:
the current code does not fetch the zip of the repo, it only fetch a zip inside the repo that contains only necessary files.

linsui commented on pkgs/by-name/do/dove/package.nix:
```diff
@@ -0,0 +1,43 @@
+    stripRoot = false;
+    # general.config.filename is hard-coded to mozilla.cfg in wrapFirefox, so removing this setting:
+    postFetch = ''
+      sed -i '/general.config.filename/d' $out/dove.cfg $out/prefs/dove.js
```
I can't find this.

jbgi replied:
it's here: https://codeberg.org/celenity/Dove/src/commit/fe9dd115f7811aadc6a8135621c621a3c4e10cdc/prefs/dove.js#L2928

linsui commented on pkgs/by-name/do/dove/package.nix:
```diff
@@ -0,0 +1,43 @@
+  meta = {
+    description = "A suite of configurations & advanced modifications for Mozilla Thunderbird with a focus on privacy, security, freedom, & usability";
+    homepage = "https://dove.celenity.dev/";
+    license = lib.licenses.gpl3Plus;
```
Why is this gpl3Plus instead of gpl3Only?

jbgi replied:
Thoward the end of the license there is is [this mention](https://codeberg.org/celenity/Dove/src/branch/pages/LICENSE#L214):

>    Phoenix
>    Copyright (C) 2024  Magnesium1062
>
>    This program is free software: you can redistribute it and/or modify it under the terms of the GNU General  Public License as published by the Free Software Foundation, either version 3 of the License, **or (at your option) any later version**.

linsui replied:
After the

> END OF TERMS AND CONDITIONS

line? The GPL license itself should not be modified as this. The license header should be added in the source code instead.

jbgi replied:
yeah, it's probably not orthodox, for me the intent is GPL3+ but I'm not a lawyer. I can open an issue to get them add proper headers.

jbgi replied:
Will be fixed in the release.

jbgi replied:
Fixed!

mweinelt commented on pkgs/by-name/tt/tts/package.nix.orig:
```diff
@@ -0,0 +1,210 @@
+  fetchFromGitHub,
+  espeak-ng,
+  tts,
+<<<<<<< HEAD
+  addBinToPathHook,
+  writableTmpDirAsHomeHook,
+=======
+  nix-update-script,
+>>>>>>> f557e19b553b (tts: add generic nix-update-script)
```
Looks like a conflict.

danth replied:
Looks like this file wasn't meant to be part of the PR at all? Note the `.nix.orig` extension

jbgi replied:
oups, right, sry about that.

jbgi commented:
> Could the config you written for Phoenix be directly tested? I am new to Nix, and I just installed NixOS. I am willing to try the changes you proposed.
> 
> Adding them to `/etc/nixos/configuration.nix` and rebuilding is enough to test them?

Assuming you have the commits of this PR in your nixpkgs then yes, the config above should work.

VeryTastyTomato commented:
In my nixpkgs? What do you mean?

jbgi commented:
> In my nixpkgs? What do you mean?

Sorry, not obvious if you are new to nixos, I meant building your nixos machine from a nixpkgs fork that would include this commit. This usually meant to migrate you config from using a nixpkgs channel to using a flake input : https://phip1611.de/blog/migrate-stock-nixos-configuration-to-flake/

But if your are using the out-of-the-box `configuration.nix` you can try phoenix (and dove) by adding a file `phoenix.nix` alongside it, with this content:
```nix
{
  pkgs,
  ...
}:
let
  version = "2025.02.21.1";
  phoenix = pkgs.fetchzip {
    url = "https://codeberg.org/celenity/Phoenix/raw/tag/${version}/archives/phoenix.zip";
    hash = "sha256-YFziMQiw0xAcMfjHQacz42zYDg9OihordfD7kRqYgPo=";
    stripRoot = false;
    # general.config.filename is hard-coded to mozilla.cfg in wrapFirefox, so removing this setting:
    postFetch = ''
      sed -i '/general.config.filename/d' $out/phoenix.cfg $out/prefs/phoenix-desktop.js
    '';
  };
  dove = pkgs.fetchzip {
    url = "https://codeberg.org/celenity/Dove/raw/tag/${version}/archives/dove.zip";
    hash = "sha256-wlFC9TMlMqMYCTFSiJ0JUUKi3hd6vRcNRovkAh3BSps=";
    stripRoot = false;
    # general.config.filename is hard-coded to mozilla.cfg in wrapFirefox, so removing this setting:
    postFetch = ''
      sed -i '/general.config.filename/d' $out/dove.cfg $out/prefs/dove.js
    '';
  };
in
{
  environment.etc."firefox/defaults/pref/phoenix-desktop.js".source =
    "${phoenix}/prefs/phoenix-desktop.js";
  environment.etc."firefox/phoenix/userjs/linux".source = "${phoenix}/userjs";
  environment.etc."firefox/phoenix/configs".source = "${phoenix}/configs";

  environment.etc."thunderbird/defaults/pref/dove.js".source = "${dove}/prefs/dove.js";

  nixpkgs.overlays = [
    (self: super: {
      # Alternatively, use programs.firefox options, but this way also works for home-manager:
      firefox = super.firefox.override {
        extraPoliciesFiles = [ "${phoenix}/policies.json" ];
        extraPrefsFiles = [ "${phoenix}/phoenix.cfg" ];
      };

      thunderbird = super.thunderbird.override {
        extraPoliciesFiles = [ "${dove}/policies.json" ];
        extraPrefsFiles = [ "${dove}/dove.cfg" ];
      };
    })
  ];
}
```
The import it from your `configuration.nix`, eg.:
```nix
  imports = [
    ./hardware-configuration.nix
    ./phoenix.nix
  ];
```

VeryTastyTomato commented:
I tried your suggestion (for Phoenix only) and it seems to work, thanks!

ddogfoodd reviewed (approved):
I wasn't sure how to test this either, so I used the method posted in https://github.com/NixOS/nixpkgs/pull/380398#issuecomment-2676715218.

Worked for me, thanks!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5684

jbgi commented:
Sorry was away (still am) from computer.
Thing is Phoenix and Dove repos now provide flakes and nixos modules (which do build from source), so not sure if this PR is still useful, especially given the relatively fast pace of releases.
